satellite/gracefulexit/endpoint.go

```diff
@@ -398,43 +412,103 @@ func (endpoint *Endpoint) processIncomplete(ctx context.Context, stream processS
+	transferQueueItem, err := endpoint.db.GetTransferQueueItem(ctx, exitingNodeID, transfer.path)
+	if err != nil {
+		return Error.Wrap(err)
 	}
 
-	transferQueueItem, err := endpoint.db.GetTransferQueueItem(ctx, nodeID, transfer.path)
+	failedCount := 0
+
+	originalOrderLimit := message.Succeeded.GetOriginalOrderLimit()
+	if originalOrderLimit == nil {
+		failedCount++
+		transferQueueItem.FailedCount = &failedCount
+		updateErr := endpoint.db.UpdateTransferQueueItem(ctx, *transferQueueItem)
+		return Error.Wrap(errs.Combine(updateErr, Error.New("Original order limit cannot be nil.")))
+	}
+	originalPieceHash := message.Succeeded.GetOriginalPieceHash()
+	if originalPieceHash == nil {
+		failedCount++
+		transferQueueItem.FailedCount = &failedCount
+		updateErr := endpoint.db.UpdateTransferQueueItem(ctx, *transferQueueItem)
+		return Error.Wrap(errs.Combine(updateErr, Error.New("Original piece hash cannot be nil.")))
+	}
+	replacementPieceHash := message.Succeeded.GetReplacementPieceHash()
+	if replacementPieceHash == nil {
+		failedCount++
+		transferQueueItem.FailedCount = &failedCount
+		updateErr := endpoint.db.UpdateTransferQueueItem(ctx, *transferQueueItem)
+		return Error.Wrap(errs.Combine(updateErr, Error.New("Replacement piece hash cannot be nil.")))
+	}
+
+	// verify that the satellite signed the original order limit
+	err = endpoint.orders.VerifyOrderLimitSignature(ctx, message.Succeeded.GetOriginalOrderLimit())
+	if err != nil {
+		failedCount++
+		transferQueueItem.FailedCount = &failedCount
+		updateErr := endpoint.db.UpdateTransferQueueItem(ctx, *transferQueueItem)
+		return Error.Wrap(errs.Combine(updateErr, err))
+	}
+
+	// verify that the original piece hash and replacement piece hash match
+	if !bytes.Equal(originalPieceHash.Hash, replacementPieceHash.Hash) {
+		failedCount++
+		transferQueueItem.FailedCount = &failedCount
+		updateErr := endpoint.db.UpdateTransferQueueItem(ctx, *transferQueueItem)
+		return Error.Wrap(errs.Combine(updateErr, ErrorHashMismatch))
+	}
+
+	// verify that the public key on the order limit signed the original piece hash
+	err = signing.VerifyUplinkPieceHashSignature(ctx, originalOrderLimit.UplinkPublicKey, originalPieceHash)
+	if err != nil {
+		failedCount++
+		transferQueueItem.FailedCount = &failedCount
+		updateErr := endpoint.db.UpdateTransferQueueItem(ctx, *transferQueueItem)
+		return Error.Wrap(errs.Combine(updateErr, err))
+	}
+
+	// TODO add nil checks/figure out a better way to get the receiving node ID
+	receivingNodeID := transfer.satelliteMessage.GetTransferPiece().GetAddressedOrderLimit().GetLimit().StorageNodeId
+
+	// get peerID and signee for new storage node
+	peerID, err := endpoint.peerIdentities.Get(ctx, receivingNodeID)
 	if err != nil {
 		return Error.Wrap(err)
 	}
+	signee := signing.SigneeFromPeerIdentity(peerID)
+
+	// verify that the new node signed the replacement piece hash
+	err = signing.VerifyPieceHashSignature(ctx, signee, replacementPieceHash)
```

Comment:
There seems to be missing `piece id` verification between replacement piece hash and replacement order limit. Would it be essential?

Reply:
It wouldn't hurt to have but it's not essential. We derive the piece ID for the replacement order limit the same way Nat did above for `calculatedNewPieceID`.

Reply:
Yes, but the exiting storage-node could provide a previously submitted piece hash.

Reply:
We get `receivingNodeID` from the replacement order limit, and I think if the receiving node has signed the hash, and the piece ID on the hash matches what we calculate it should be, everything should be okay.
But I could be overlooking something and it wouldn't hurt to also check that the piece ID of the order limit is the same as the piece ID of the hash.

Reply:
If I send you a properly signed replacement piece hash for a wrong piece id, what would happen currently?

Reply:
This check above should fail
```
calculatedNewPieceID := transfer.rootPieceID.Derive(receivingNodeID, transfer.pieceNum)
if calculatedNewPieceID != replacementPieceHash.PieceId {
	return ErrInvalidReplacementPieceID
}
```